Use traces to extract the input for `settlement::Observer`

sunce86 opened the pull request:
# Description
Implements first part of https://github.com/cowprotocol/services/issues/3177

The expectation for "auction_id" remain the same: `auction_id` is expected at the end of the `/settle` call/sub_call.

# Changes

- [ ] Uses `trace_transaction` function to get traces and extract the calldata of `settle` call

## How to test
Existing tests show that no regression bugs were introduced.

Final e2e test should cover this part of the code. Although by manually testing the `trace_transaction` I am pretty sure the code is correct.
<!--
## Related Issues

Fixes #
-->

sunce86 commented on crates/autopilot/src/domain/settlement/mod.rs:
```diff
@@ -29,8 +29,6 @@ pub struct Settlement {
     /// The effective gas price of the settlement transaction.
     gas_price: eth::EffectiveGasPrice,
-    /// The address of the solver that submitted the settlement transaction.
-    solver: eth::Address,
```
Not strictly related to the task but removed for the sake of completeness so we don't have to rethink how this field should be populated.

squadgazzz commented on crates/autopilot/src/domain/settlement/transaction/mod.rs:
```diff
@@ -30,23 +31,28 @@ impl Transaction {
     pub fn new(
         transaction: &eth::Transaction,
         domain_separator: &eth::DomainSeparator,
+        settlement_contract: eth::Address,
     ) -> Result<Self, Error> {
```
Not related, but it looks like not all the constructors returning `Result` were renamed.

squadgazzz commented on crates/autopilot/src/domain/settlement/transaction/mod.rs:
```diff
@@ -116,6 +122,18 @@ impl Transaction {
+                .interface
+                .abi
+                .function("settle")
+                .unwrap()
```
Any way to avoid `unwrap()` call here?

sunce86 replied:
This will never fail 🎉

MartinquaXD reviewed (approved):
Logic looks good to me.
The bit where the auction id gets spliced off the `settle()` call might have to change in the future depending on the helper contracts. If the auction id gets appended to the overall function call the solidity logic of the helper contracts doesn't have to handle it at all which would probably result in gas savings. Not sure if those are big enough to matter in practice, though.

mstrug commented on crates/autopilot/src/domain/settlement/transaction/mod.rs:
```diff
@@ -30,23 +31,28 @@ impl Transaction {
-            .input
-            .0
-            .split_at(transaction.input.0.len() - META_DATA_LEN);
+        let (data, metadata) = calldata.0.split_at(calldata.0.len() - META_DATA_LEN);
```
This can panic if `calldata` is shorter than `META_DATA_LEN`. I'm not sure if this can happen, but we could use `split_at_checked()` function.

sunce86 replied:
Ended up refactoring the error handling a bit, to notify more loudly if there is an execution bug.

mstrug commented on crates/autopilot/src/infra/blockchain/mod.rs:
```diff
@@ -151,6 +153,16 @@ fn into_domain(
+        trace_calls: traces
+            .into_iter()
+            .filter_map(|trace| match trace.action {
+                web3::types::Action::Call(call) => Some(eth::TraceCall {
```
The conversion could be moved to `TraceCall` `From<web3::types::Call>` trait implementation.

sunce86 replied:
would like to keep it as is since it's used only once

sunce86 commented:
Refactored the error handling to notify more loudly if execution client bug is detected.

Also, added the unit test to verify that `transfer` calls that emit settlement event are still properly handled: https://etherscan.io/tx/0x030623e438f28446329d8f4ff84db897907fcac59b9943b31b7be66f23c877af
